feat(extensions):  storidge evolution

xAt0mZ opened the pull request:
Evolution of storidge extension.

xAt0mZ commented:
Feedback on API:
1. GET /nodes/{node-id}
* All fields of `properties` have a ':' at then end. So we have keys like `domain:` instead of `domain`.
2. GET /volumes/{volume-id}
* On a same point, we have a field here that is `allocate%`, the `%` needs to be removed.
* Following the specs, we should have 26 fields available on this request, we only have 14 on requests, the missing ones are the ones commented [here](https://github.com/portainer/portainer/blob/9957c1debe4f555465a47d7c5219761da585b82f/app/extensions/storidge/models/volume.js)

xAt0mZ commented:
Small info on volume update, trying to update labels randomly ends on an error:

<details>
 <summary>UPDATE 1: success</summary>

![2019-02-26-195104_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/53438356-06341b80-3a00-11e9-8724-887585150f1f.png)
![2019-02-26-195147_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/53438355-06341b80-3a00-11e9-91bb-0fb10607fbff.png)

</details>

<details>
 <summary>UPDATE 2: success</summary>

![2019-02-26-195150_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/53438354-06341b80-3a00-11e9-867d-8d326378f50f.png)
![2019-02-26-195058_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/53438359-06341b80-3a00-11e9-88bc-6ba1e0f000e9.png)

</details>

<details>
 <summary>UPDATE 3: error</summary>

![2019-02-26-195048_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/53438361-06341b80-3a00-11e9-89eb-967322445723.png)
![2019-02-26-195201_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/53438353-059b8500-3a00-11e9-85e8-42fb4c8fccc9.png)

</details>

xAt0mZ commented:
Not working requests: 
* `POST /volumes/{volume-id}/snapshot`
returns with `volume-id` = 1 or = portainer (respectively existing volume id and name).
```
{
    "content": "Volume was not found"
}
```

* `GET /volumes/{volume-id}/snapshots`
returns with a valid `volume-id` (vdisk id).
```
{
    "content": "Usage: cio snapshot COMMAND [options] [YYYY-MM-DD-HHMM-NODEID-VDID]\n\nManually create a snapshot, print info, list all snapshots or remove a snapshot\n\nCommands:\n    create    Create snapshot of a volume\n    help      Show command usage information\n    info      Display snapshot information by snapshot id\n    list      List snapshots on a volume\n    remove    Remove snapshot by snapshot id\n\nOptions:\n    -v | --volume <volume name>         apply to volume identified by volume name\n    -V | --vdisk <id>                   apply to volume identified by vdisk ID\n    YYYY-MM-DD-HHMM-NODEID-VDID         snapshot ID, present only for remove"
}
```

Doesn't work either with volume name as `volume-id` :
```
{
    "content": "Fail: Volume vdportainer not found"
}
```

deviantony commented:
First version ready for testing.

xAt0mZ commented:
On volume update:
There is no way to remove labels when trying to update a volume, as only transmitted data are taked into account on POST ; you can't remove a field with this behaviour, you can only add new fields (previously empty ones) or modify existing ones.

deviantony commented:
@xAt0mZ you should be able to access the latest version of the Storidge API by rebuilding entirely your Vagrant environment.

Here is the list of things we need to add/update for the next version of the Storidge integration.

## Additions

- [ ] Add support for the rescan features via a `Rescan drives` button in the drive list view

![Portainer (41)](https://user-images.githubusercontent.com/5485061/54174287-912a1100-44ea-11e9-9f10-e77c587c858f.png)

Based on the following API endpoint:

```
POST /drives/rescan
```

## Enhancements

- [ ] Add a new `Add a node` button in the node datatable panel that will display the current information text underneath it after clicking on it

![Portainer (40)](https://user-images.githubusercontent.com/5485061/54174363-fed63d00-44ea-11e9-932c-5d3b3732e8ab.png)

- [ ] Remove the ability to select nodes via checkboxes from the nodes datatable component and relocate the following action buttons to the node details view: `Remove node`, `Enter maintenance mode` and `Exit maintenance mode`.

![Portainer (38)](https://user-images.githubusercontent.com/5485061/54174395-1ad9de80-44eb-11e9-9b92-d7b746a275b8.png)

Note that the `Enter maintenance mode` button should be disabled if the status of the node is 'maintenance'. The `Exit maintenance mode` button should be disabled if the status of the node is 'normal'.

- [ ] Update the icon color for the node status in the node datatables to yellow/orange when status is set to 'maintenance' (see comment in picture above)

-  [ ] Update the UX for the snapshot interval property in the profile creation to support a blank value. The suggestion here is to add an extra switch labelled `Enable periodic snapshots` right under the `Enable snapshot` switch to define a value for the snapshots. If the switch is not enabled, the snapshot interval property will be blank.

- [ ] Update the UX for the snapshot interval property selection to use minutes instead of seconds

- [ ] Replace the icon color of the cluster status label to red when the status is set to `Alert`

![Portainer (39)](https://user-images.githubusercontent.com/5485061/54174564-fcc0ae00-44eb-11e9-9ae3-91d28bb2d73b.png)

## Bugfixes

- [ ] Fix a bug preventing the removal of a drive

```
The DELETE /drives/{drive-id} popup did not work in Portainer. However Postman was able to delete the drive. Please check Postman collection example.
```

- [ ] Fix a bug preventing the addition of a drive

```
 The POST /drives returns “Failure, unable to add Storidge drive”. However Postman was able to add the drive so maybe the request body contents is wrong? Please check Postman collection example.
```

- [ ] Fix a bug causing an invalid request when trying to change the capacity of a volume

```
POST /volumes/{volume name|vid} to change capacity failed because of bad parameter, e.g. API sent `cio volume update -v foo --capacity 2000GB` when it should send `cio volume update -v foo --capacity 2000`, i.e. no ‘GB’.
```

- [ ] Fix a bug causing an invalid request when trying to change a volume IOPS

```
 POST /volumes/{volume name|vid} to change IOPS failed because of bad parameters, API sent `cio volume update -v foo --iopsmin %!s(float64=1000) --iopsmax %!s(float64=3000)` when it should send `cio volume update -v foo --iopsmin 1000 --iopsmax 3000`
```

xAt0mZ commented:
> Update the UX for the snapshot interval property in the profile creation to support a blank value. The suggestion here is to add an extra switch labelled Enable periodic snapshots right under the Enable snapshot switch to define a value for the snapshots. If the switch is not enabled, the snapshot interval property will be blank.

What about Max Snapshots ? Same behaviour ?

xAt0mZ commented:
> Fix a bug preventing the removal of a drive

> Fix a bug preventing the addition of a drive

Can't test those ones as I don't have the setup for it.

> Fix a bug causing an invalid request when trying to change the capacity of a volume

Storidge API is returning a value like `20GB` on query but is waiting for an integer on update (by testing, waiting for a size in GB) so the field `20GB` is invalid. One should write `21` in the field to increase the volume capacity to 21GB.
Tested with this (left the form content, right the associate generated payload)
![2019-03-12-185917_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/54224093-1caa9e80-44f9-11e9-93f6-309663cc26d3.png)

and it is working fine
![2019-03-12-190248_935x799_scrot](https://user-images.githubusercontent.com/7642331/54224275-78752780-44f9-11e9-9a21-14db2901e000.png)


Storidge needs to fix this behaviour by either returning a value without GB/TB/whatever on query, or wait for a string like `21GB` and handle themselves the conversion to a real volume for their system.

deviantony commented:
New update:

# Enhancements

- [ ] Disable volume snapshot retrieval when snapshot is not enabled for the associated volumes. Do not display the snapshots panel in that case.

# Bugfixes

- [ ] Fix the pagination on the drive list view to remember selected pagination filter on refresh

- [ ] Ensure that the pagination filter is working correctly in all the views

- [ ] Fix a bug associated to the "Add to storage pool" button

Request body seems to be the following:

```json
{"device":"none"}
```

Whereas it should be:

```json
{
  "device": "/dev/sdr",
  "node": "t1"
}
```

deviantony commented:
New update

# Enhancements

- [ ] Show capacity details in units of TB for node details view. Currently showing in unit of bytes

- [ ] In the node details view: Align the values in right column to same indent. Currently sections 1 and 4 are aligned but 2 and 3 are at different offsets

# Bugfixes

- [ ] Remove drive function not working. No request issued to REST API from UI.

- [ ] Empty node details view (happens with latest version connected to my local Vagrant environment, note sure Storidge reported that one)

![Portainer (42)](https://user-images.githubusercontent.com/5485061/54894168-377c0a80-4f1e-11e9-8f6d-64852734f58e.png)

- [ ] In the node details view, invalid state checking to disable/enable the Exit maintenance mode button.

> Exit maintenance mode] button is not active on a cordoned node, so cannot uncordon node from UI. Check node status and activate maintenance buttons accordingly:

> If node status is cordoned, then activate [Exit maintenance mode] button. Inactivate [Remove node] and [Enter maintenance mode] buttons

> If node status is normal, activate [Enter maintenance mode] and [Remove node] buttons, and inactivate [Exit maintenance mode] button

- [ ] In the node details view, When node status is cordoned, the Node details view shows 0 values for most parameters which is correct. However the Bandwidth details section has non-zero values

- [ ] Fix a bug when creating a profile

> [Create the profile] button returns a 500 error code {"error":"json: cannot unmarshal array into Go struct field Profile.label of type map[string]string"}

> Request body was: 

```json
{"name":"SNAP","capacity":20,"directory":"/cio/SNAP","provision":"thin","type":"ssd","level":2,"iops":{"min":100,"max":2000},"bandwidth":{"min":1,"max":100},"filesystem":"btrfs","service":{"snapshot":{"enabled":false},"encryption":{"enabled":false}},"interface":{"type":"nfs","driver":"","network":"","conf":""},"label":[]}
```

> The Interface parameters do not have any values. Suggestion is to not send the Interface related parameters when the profile does not specify a Interface.

- [ ] When creating a profile, if user enables snapshot the UI sends fields ‘interval’, ‘max’ and ‘start’ as part of the profile. When using the profile to then create a volume, volume creation fails if the ‘interval’ value is 0 since cio considers that an invalid value.

> Suggest changing the layout of the Create a profile view to:

![unnamed](https://user-images.githubusercontent.com/5485061/54894292-cf79f400-4f1e-11e9-87a9-54ad29aa37ec.jpg)

> If user enable snapshots, show the snapshot max settings and default snapshot max count to 1. Send only the ‘enabled: yes’ and ‘max: X’ fields in the profile if periodic snapshots are NOT enabled.

> If user enable periodic snapshots, show the snapshot interval and default snapshot interval to 1440, i.e. one snapshot per day. Send fields ‘enabled: yes’, ‘max: X’, ‘interval: Y’ in the profile. Do NOT send field ‘start’ as there is no setting in the UI for user. Cio can support a start time for snapshot based on user request later.

Note: Snapshot interval and snapshot max with their selectors should be on a separate line as it is done right now instead of closed to the switch as it is specified in the screen above.

- [ ] Currently when the user sets the snapshot interval to 2, in the created profile the value shows as 120, i.e. UI is treating the interval input as hours instead of minutes.

xAt0mZ commented:
> In the node details view: Align the values in right column to same indent. Currently sections 1 and 4 are aligned but 2 and 3 are at different offsets

Can't figure how the padding is made, prob need to rework a bit how style is done here.

> In the node details view, invalid state checking to disable/enable the Exit maintenance mode button.

I need to have the exact Status values to fix that.

deviantony commented:
Based on Storidge feedback for Node status and maintenance mode:

> If node status is cordoned, then activate [Exit maintenance mode] button. Inactivate [Remove node] and [Enter maintenance mode] buttons

> If node status is normal, activate [Enter maintenance mode] and [Remove node] buttons, and inactivate [Exit maintenance mode] button

Are these statuses available in the API ?

xAt0mZ commented:
1. > Are these statuses available in the API ?

The status is given by the API when retrieving the node details, but I don't know the Status value when node is under maintenance (need to modifiy [this](https://github.com/portainer/portainer/pull/2711/files#diff-25714b9c1f30857a2af231b674647bce) and [this](https://github.com/portainer/portainer/pull/2711/files#diff-3386a7740ffe8c5a226fd0c5471fa1d7R28) files with the real `maintenance` status).
I can't cordon a node with my dev env to know the exact status value returned by the API. Changed it to `cordoned` as it seems to be a consistent value for the cordoned / maintenance state.

**Is it possible to have a full list of possible statuses for nodes, drives and cluster to manage all scenarios ?**

2. > In create profile view, Inconsistent font for entries: currently bold for directory and capacity and normal for the rest.

`rdash-ui` css is overriding `bootstrap` css for `select` tags . Don't know how to fix this atm (maybe by importing rdash-ui before bootstrap when bundling `vendor.vendor.css` file) 
![2019-03-27-160044_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/55086931-886c3a00-50a9-11e9-8df3-1c4158fed453.png)

3. > Empty node details view (happens with latest version connected to my local Vagrant environment, note sure Storidge reported that one)

Related to NodeDetailsModel [refactor](https://github.com/portainer/portainer/pull/2711/commits/cd6587ebcdca1a7be5482e48654859f6878d57c6#diff-14852cd1921de42eac66f9ccb372cc34) for latest storidge version. Need to rebuild Storidge vagrant setup to fix this.

4. > In the node details view, When node status is cordoned, the Node details view shows 0 values for most parameters which is correct. However the Bandwidth details section has non-zero values

We are displaying the data retrieved by the API without any modification nor default values (see [StoridgeNodeService](https://github.com/portainer/portainer/blob/feat-evol-storidge/app/extensions/storidge/services/nodeService.js#L31) and the associated [Model](https://github.com/portainer/portainer/blob/feat-evol-storidge/app/extensions/storidge/models/node.js#L8) )

5. > In Volume details view, the snapshot list is not displaying. However browser shows the response body returning the snapshot list.

The data returned by the API don't match the specs given on the doc
Specs : (array)
![2019-03-27-174625_448x509_scrot](https://user-images.githubusercontent.com/7642331/55095696-df2d4000-50b8-11e9-8cf8-252fa6c31780.png)
Actual response : (object with 1 field = 1 item)
![2019-03-27-174954_540x202_scrot](https://user-images.githubusercontent.com/7642331/55095726-ece2c580-50b8-11e9-8ccd-a2edeaef25a6.png)

Refactored the code to match the current API response format. Will it be modified to match the specs at a time or will it stay like this ?

deviantony commented:
> In create profile view, Inconsistent font for entries: currently bold for directory and capacity and normal for the rest.

That's a global UI matter, we'll remove it from the state of work.

deviantony commented:
New update.

# Enhancements

- [ ] In the node details view: Align the values in right column to same indent. Currently sections 1 and 4 are aligned but 2 and 3 are at different offsets

- [ ] In Volume details/Storidge details section, remove ‘Volume’ and ‘Replication’. ‘Volume’ is redundant since ‘Name’ exists. ‘Replication’ feature is not available yet.

- [ ] In the node details view, Align the values in right column to same offset. Currently the values of Node details, Bandwidth details, Capacity details and IOPS details are at different offsets

- [ ] In Create profile/Profile configuration section, change ‘Interface driver’ to ‘Network driver’. Change ‘Interface network’ to ‘Network name’.

- Bugfixes

- [ ] In snapshot creation, The API is not picking up the ‘Description’ of the snapshot sent by UI. See format of request body from UI.

![unnamed (1)](https://user-images.githubusercontent.com/5485061/55126730-cc774180-5172-11e9-912e-91bc5608416d.jpg)

Use example below for the request body:

```json
{
        "opts": {
                        "description": "Manual snapshot on volume rotate"
        }

}
```

- [ ] In the node details view, Cannot exit maintenance mode from UI since [Exit maintenance mode] button is not active on a cordoned node. UI should activate the remove and maintenance buttons based on node status.

> Node status will be ‘normal’, ‘cordoned’ or ‘leaving. ‘normal’ means part of the cluster and operating. ‘cordoned’ means part of cluster but in maintenance mode. ‘leaving’ means in the process of being removed from the cluster.

> Set rules as:
If node status is ‘cordoned’, then activate [Exit maintenance mode] button. Inactivate [Remove node] and [Enter maintenance mode] buttons
If node status is ‘normal’, activate [Enter maintenance mode] and [Remove node] buttons, and inactivate [Exit maintenance mode] button
If node status is ‘leaving’, inactivate [Enter maintenance mode], [Remove node], [Exit maintenance mode] buttons

- [ ] In the profile creation view, When [Enable snapshots] button is true, the profile sent to API has fields ‘interval’ and ‘start’. ‘interval’ is used only with periodic snapshots so should not be sent when [Enable periodic snapshots] is false. ‘start’ is not currently supported as a feature so should not be sent in the profile. The only fields to send when [Enable snapshots] button is true are ‘snapshot:’, ‘enabled: yes’ and ‘max: Y’.

- [ ] In the profile creation view, Note that when [Enable snapshots] button is false, the ‘snapshot:’, ‘enabled: no’, ‘interval: 60’, ‘max: 10’ and ‘start: 1440’ key values are sent. Since snapshot is disabled this does not cause a problem in creating volumes. However if user enable snapshots later but not periodic snapshots, then creating volumes with the profile will fail. Suggestion is to only send ‘snapshot:’, ‘enabled: no’, and ‘max: 10’ key values only when [Enable snapshots] if false. Do not send fields ‘interval: 60’ and ‘start: 1440’. These fields are sent only if the user [Enable periodic snapshots].

- [ ] In the profile creation/profile details views, UI defaults the directory path to follow name of profile. If profile name is set to upper case, directory path name is also upper case, e.g. Profile name = TEST, then directory path = /cio/TEST. Always set directory path to lower case regardless of profile name is upper case or lower case.

xAt0mZ commented:
> In the profile creation view, When [Enable snapshots] button is true, the profile sent to API has fields ‘interval’ and ‘start’. ‘interval’ is used only with periodic snapshots so should not be sent when [Enable periodic snapshots] is false. ‘start’ is not currently supported as a feature so should not be sent in the profile. The only fields to send when [Enable snapshots] button is true are ‘snapshot:’, ‘enabled: yes’ and ‘max: Y’.

> In the profile creation view, Note that when [Enable snapshots] button is false, the ‘snapshot:’, ‘enabled: no’, ‘interval: 60’, ‘max: 10’ and ‘start: 1440’ key values are sent. Since snapshot is disabled this does not cause a problem in creating volumes. However if user enable snapshots later but not periodic snapshots, then creating volumes with the profile will fail. Suggestion is to only send ‘snapshot:’, ‘enabled: no’, and ‘max: 10’ key values only when [Enable snapshots] if false. Do not send fields ‘interval: 60’ and ‘start: 1440’. These fields are sent only if the user [Enable periodic snapshots].

I don't get how they are retrieving these values. We can't send a `start` as we don't have any field like this in our code.

Here are the payloads sent when creating profiles without snapshots.
![2019-03-28-161044_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/55171375-88873b00-5178-11e9-8114-5651b3639287.png)

With snapshots without periodics
![2019-03-28-161109_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/55171374-88873b00-5178-11e9-9faf-2c5858e55cd4.png)

With periodics
![2019-03-28-161130_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/55171373-88873b00-5178-11e9-93ba-31149c435087.png)

xAt0mZ commented:
Updated `app.css` file to fix the select style as `rdash-ui` is forcing the following style
```css
@media screen and (-webkit-min-device-pixel-ratio: 0)
select {
    font-family: Arial, Helvetica, sans-serif;
}
```
So it doesn't respect the principle of having 1 font family on the app.

Modification is replicated everywhere on the app, so some select elements may appear "broken" on small screen sizes.

Before
![2019-03-28-181619_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/55178806-67c5e200-5186-11e9-8b7a-697e72ef0d74.png)

After
![2019-03-28-181625_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/55178805-672d4b80-5186-11e9-8e4b-4cca15b7b5b7.png)
![2019-03-28-182143_1920x1080_scrot](https://user-images.githubusercontent.com/7642331/55178804-672d4b80-5186-11e9-8a34-a30db41e4b8b.png)

deviantony commented:
# v6

New update

# Enhancements



# Bugfixes

- [ ]  GET /nodes/<node> all buttons are always disabled whether node is in cordoned or uncordoned state

- [ ] In Volume details/Storidge details view, the allocated, IOPS Min, IOPS Max, Local Drive Only, Drive Type, Max Snapshots fields do not show values since the JSON response is now returning the keys in camel case.

- [ ] POST /profiles – Create profile/Modify defaults view always has the Interface type set to ‘nfs’. Default the Interface type to ‘null’ value in the UI. This way the ‘Interface:’ and ‘type: nfs’ key value will NOT be sent in profile by default. The cio software is rejecting the profile during volume creation because other Interface fields are missing.

- [ ]

deviantony commented:
* In Storidge drives view, the [Remove] and [Add to storage pool] button allows multiple drives to be selected for addition or removal. The cioctl drive command however only operates on one drive at a time.

This is now fixed, there is no way to select nodes from the node table anymore, you can use the remove/add to storage pool function from the node details view.

* GET /nodes/ all buttons are always disabled whether node is in cordoned or uncordoned state

Fixed, see my comment below for more details about this and a potential error.

* In Volume details/Storidge details view, the allocated, IOPS Min, IOPS Max, Local Drive Only, Drive Type, Max Snapshots fields do not show values since the JSON response is now returning the keys in camel case.

Fixed.

* POST /profiles – Create profile/Modify defaults view always has the Interface type set to ‘nfs’. Default the Interface type to ‘null’ value in the UI. This way the ‘Interface:’ and ‘type: nfs’ key value will NOT be sent in profile by default. The cio software is rejecting the profile during volume creation because other Interface fields are missing.

Interface type now defaults to null and Portainer will not send the interface.type property when not specified.

* POST /profiles – Create profile/Modify defaults view, if Enable periodic snapshots button is off, do not send the ‘interval: 0’ and ‘start: 0’ fields. Interval 0 is an invalid value and will be rejected by cio software during volume creation. The start field is not a supported feature and should not be in the profile.

I was not able to reproduce that problem. The 'start' property is never sent and the 'interval' property is only sent when the "Enable periodic snapshots" switch is on.

* POST /profiles – Create profile/Modify defaults view the Limit bandwidth fields are not sent in the profiles. Regardless of actual values set, the profile ends up with default IOPS values.

This is pretty strange as the `bandwidth.max` and `bandwith.min` options are actually passed in the body of the request for the profile creation. It seems that they are ignored by the API though, see request body example below:

```
bandwidth: {min: 7, max: 77}
max: 77
min: 7
capacity: 20
directory: "/cio/test07"
filesystem: "btrfs"
interface: {driver: "", network: "", conf: ""}
conf: ""
driver: ""
network: ""
label: {}
level: 2
name: "test07"
provision: "thin"
service: {snapshot: {enabled: false}, encryption: {enabled: false}}
encryption: {enabled: false}
enabled: false
snapshot: {enabled: false}
enabled: false
type: "ssd"
```

During my tests I encoutered the following problem:

* Issue with remove node action

After hitting the remove node action, the API responded with a 409 and the following response content:

{"content":"perl: warning: Setting locale failed.\nperl: warning: Please check that your locale settings:\n\tLANGUAGE = (unset),\n\tLC_ALL = (unset),\n\tLC_CTYPE = \"fr_FR.UTF-8\",\n\tLANG = \"en_US.UTF-8\"\n    are supported and installed on your system.\nperl: warning: Falling back to a fallback locale (\"en_US.UTF-8\").\nSucceed: node 3fdc49af state is remove"}

It seems that the node was removed successfully though, as its status changed to leaving after that. I recommend fixing the return code and response content.

* Issue with node status

I've created a 4 node clusters and put node 3 in maintenance mode with success. The nodes tables (GET /nodes/) reports the node status as "cordoned" but the node details (GET /nodes/storidge-3) reports the node status as "Normal", causing inconsistencies in the UI (e.g. not able to exit maintenance mode as node status is considered normal and not cordoned).

Response from GET /nodes/:

```
{"nodes":{"storidge-1":{"ip":"10.0.9.10","id":"bfe21a39","role":"sds","status":"normal"},"storidge-2":{"ip":"10.0.9.11","id":"4f905ef5","role":"backup1","status":"normal"},"storidge-3":{"ip":"10.0.9.12","id":"3fdc49af","role":"backup2","status":"cordoned"},"storidge-4":{"ip":"10.0.9.13","id":"6e529ced","role":"standard","status":"normal"}}}
```

Response from GET /nodes/storidge-3:

```
{"name":"storidge-3","properties":{"domain":"bfe21a39","domainID":"bf9ad05b","nodes":"4","status":"Normal"}}
```

deviantony commented:
# v7

New update

# Enhancements

- [ ] Add the Condition field or some color coded indicator to the Node Details view

- [ ] Add the cluster Status field to the Node Details view

- [ ] in the Node Details view with node in ‘normal’ or ‘cordoned’ status, the [Remove node], [Enter maintenance mode], and [Exit maintenance mode] button are inactive.

> Note from alapenna: I think this should now leverage the new Condition field instead of the Status field.

# Bugfixes

- [ ] The [Add to storage pool] button does not show. The UI sends a GET with serial number ‘none’ and the API responds with 409 code.

Instead of issuing the GET request, suggest displaying the drive details and a [Add to storage pool] button. The ‘available’ drive information to display is the same as from the Storidge Drives view.  

- [ ] When node in cordoned mode the [Exit maintenance mode] button is inactive but [Remove node] and [Enter maintenance mode] buttons are active. This is because ‘Status’ field is confusing (see comments and my question below). Use ‘Status’ field from GET /nodes to verify cordoned status instead of using GET /nodes/<node name>

- [ ] On cio 2748, note that after a volume is created with profile that has bandwidth settings, the Volume Details view show only the iops and not the bandwidth info. The iops values are null.

The Volume Details view needs to add the Bandwidth Min and Max fields or toggle the IOPS Min and Max field with bandwidth values.

- [ ] The type field for profile interfaces is currently immutable so the default null value cannot be changed. Because of this when a profile with an interface is created, everything but the type can be set and invalid profiles like this are all that can be created: 

```
$ cio profile info b
bandwidth:
  max: 100
  min: 1
capacity: 20
directory: /cio/b
interface:
  conf: '*(rw,sync,fsid=0,secure,root_squash,all_squash)'
  driver: macvlan
  network: cionet
iops: {}
label: {}
level: 2
local: no
provision: thin
service:
  compression: no
  dedupe: no
  encryption:
    enabled: no
  replication:
    destination: none
    enabled: no
    interval: 120
    type: synchronous
  snapshot:
    enabled: yes
    interval: 1440
    max: 1
    start: 0
type: ssd
```

Recommended solution:

There should be a drop-down for interface fields that defaults to null, similar to how the network driver field is set up, but can then be changed to nfs and other interface types that might be added down the line. 

Also setting both of these drop-downs (interface type and network driver) to include blank entries at the top would improve usability as users won't have to commit if they select a value from the drop-downs then change their mind.

- [ ] I also found that creating a default profile straight from the profiles list page sends the suggested values for bandwidth and iops along. It shouldn't be sending anything for these. Please make the profile sent by clicking 'Create the profile' the same as that sent by clicking 'Modify defaults...' and making no changes. 

I've included screenshots for reference, they're named after the buttons used to create the default profiles. 

![create-the-profile](https://user-images.githubusercontent.com/5485061/56629897-4bdd2f80-66a3-11e9-9b69-44b230ba312c.png)

![modify-defaults](https://user-images.githubusercontent.com/5485061/56629899-54356a80-66a3-11e9-9f79-9862b6f14af2.png)

xAt0mZ commented:
> * [x]  Add the Condition field or some color coded indicator to the Node Details view
> * [x]  Add the cluster Status field to the Node Details view
> * [x]  in the Node Details view with node in ‘normal’ or ‘cordoned’ status, the [Remove node], [Enter maintenance mode], and [Exit maintenance mode] button are inactive.

Seems that `Condition` field is the cluster status, and `Status` is still the node status. Not sure if that is the intended behaviour.
After putting a node in maintenance, i'm getting the following data
```
{  
   "name":"storidge-4",
   "properties":{  
      "condition":"normal",
      "domain":"da344bf0",
      "domainID":"fca5841e",
      "freeBandwidth":"0.000Mi",
      "freeCapacity":"0",
      "freeIOPS":"0",
      "hdds":"0",
      "metadataVersion":"0.0",
      "nodes":"4",
      "provisionedBandwidth":"0.000Mi",
      "provisionedCapacity":"0",
      "provisionedIOPS":"0",
      "ssds":"0",
      "status":"cordoned",
      "totalBandwidth":"0.000Mi",
      "totalCapacity":"0",
      "totalIOPS":"0",
      "usedBandwidth":"0.000Mi",
      "usedCapacity":"0",
      "usedIOPS":"0",
      "vdisks":"0"
   }
}
```

xAt0mZ commented:
**Comments regarding v6**

> * In Storidge drives view, the [Remove] and [Add to storage pool] button allows multiple drives to be selected for addition or removal. The cioctl drive command however only operates on one drive at a time.

Not sure about the need to remove this part that makes easier the bulk of operations. Code was done to support this API behaviour as operations (add / remove) were done successively on each concerned drives (addition on `available` drives / deletion on `faulty` drives).

deviantony commented:
Regarding v6 update.

> Not sure about the need to remove this part that makes easier the bulk of operations. Code was done to support this API behaviour as operations (add / remove) were done successively on each concerned drives (addition on available drives / deletion on faulty drives).

This behavior was requested by Storidge.

deviantony commented:
# v8

New update.

## Enhancements

- [ ] Change the 'Cluster condition' label to 'Operating condition'

deviantony commented:
# v9

## Enhancements

- [ ] For the [Add to storage pool] button it should only be allowed select and add one drive at a time. Currently more than one available drive can be selected which will cause inconsistent results.

We should add a new column in the table (Actions) with a "Add to storage pool" button for each drive. That would provide a better UX.

xAt0mZ commented:
> * the UI will always stay on the first IP address it finds for Storidge API. Even when the node is removed from the cio cluster (not Swarm cluster), and Portainer service is then moved to a new node, the UI will keep sending requests to the IP address it discovers first.

Tested but not able to know where the problem is. I have the following config:
• 10.0.9.10 -> sds
• .11 -> backup1
• .12 -> backup2
• .13 -> standard

Removing sds from cio cluster result in Portainer (running on `sds:9000`) not able to access the storidge features.
Should go on Portainer on `backup1:9000` to be able to access the storidge features
After that, Portainer on `sds:9000` is able to access storidge features again

------
On external Portainer instance (local running), connected to cio cluster through sds, I'm not able to access storidge features.
Also, portainer docker volume (with cio:latest driver) is not showed even if it exists (can see it on `sds:9000` Portainer instance)

deviantony commented:
# v10

## Backend

- [ ] the UI will always stay on the first IP address it finds for Storidge API. Even when the node is removed from the cio cluster (not Swarm cluster), and Portainer service is then moved to a new node, the UI will keep sending requests to the IP address it discovers first.

Report from LP: https://github.com/portainer/portainer/pull/2711#issuecomment-487365810
